File: crates/runlet-core/src/event.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum EventKind {
	#[serde(rename = "job.queued")]
	JobQueued,
	#[serde(rename = "job.started")]
	JobStarted,
	#[serde(rename = "stdout")]
	Stdout,
	#[serde(rename = "stderr")]
	Stderr,
	#[serde(rename = "job.completed")]
	JobCompleted,
	#[serde(rename = "job.failed")]
	JobFailed,
	#[serde(rename = "job.canceled")]
	JobCanceled,
	#[serde(rename = "job.cancel_requested")]
	JobCancelRequested,
}

impl EventKind {
	pub fn as_str(&self) -> &'static str {
		match self {
			EventKind::JobQueued => "job.queued",
			EventKind::JobStarted => "job.started",
			EventKind::Stdout => "stdout",
			EventKind::Stderr => "stderr",
			EventKind::JobCompleted => "job.completed",
			EventKind::JobFailed => "job.failed",
			EventKind::JobCanceled => "job.canceled",
			EventKind::JobCancelRequested => "job.cancel_requested",
		}
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventRecord {
	pub seq: u64,
	pub ts: DateTime<Utc>,
	pub kind: EventKind,
	pub payload: Value,
}

#[derive(Debug, Clone)]
pub struct EventInput {
	pub ts: DateTime<Utc>,
	pub kind: EventKind,
	pub payload: Value,
}

impl EventInput {
	pub fn new(kind: EventKind, payload: Value) -> Self {
		Self {
			ts: Utc::now(),
			kind,
			payload,
		}
	}
}

File: event.go
package runlet

import "time"

type EventKind string

const (
	EventKindJobQueued          EventKind = "job.queued"
	EventKindJobStarted         EventKind = "job.started"
	EventKindStdout             EventKind = "stdout"
	EventKindStderr             EventKind = "stderr"
	EventKindJobCompleted       EventKind = "job.completed"
	EventKindJobFailed          EventKind = "job.failed"
	EventKindJobCanceled        EventKind = "job.canceled"
	EventKindJobCancelRequested EventKind = "job.cancel_requested"
)

type EventRecord struct {
	Seq     uint64         `json:"seq"`
	TS      time.Time      `json:"ts"`
	Kind    EventKind      `json:"kind"`
	Payload map[string]any `json:"payload"`
}

func (k EventKind) terminal() bool {
	switch k {
	case EventKindJobCompleted, EventKindJobFailed, EventKindJobCanceled:
		return true
	default:
		return false
	}
}

type eventInput struct {
	ts      time.Time
	kind    EventKind
	payload map[string]any
	ack     chan struct{}
}
